src/Container/Container.php

```diff
@@ -165,4 +216,58 @@ public function __call($name, $arguments)
             throw new \BadMethodCallException("No application method named {$name}()");
         }
     }
+
+    /**
+     * Instantiate an object of named class, recursively resolving dependencies
+     *
+     * @param string $className Fully qualified class name
+     * @return mixed
+     * @throws \ReflectionException
+     */
+    protected function resolveInstance(string $className)
+    {
+        $class = new \ReflectionClass($className);
+
+        if (!$class->isInstantiable()) {
+            throw new \ReflectionException(sprintf('Class %s cannot be instantiated', $className));
+        }
+
+        $parameterValues = [];
+        if (($constructor = $class->getConstructor())) {
+            $parameterValues = $this->resolveParameters(
+                $constructor->getParameters()
+            );
+        }
+
+        return $class->newInstanceArgs($parameterValues);
+    }
+
+    /**
+     * Recursively resolve function parameters using type hints
+     *
+     * @param \ReflectionParameter[]
+     * @return mixed
+     */
+    protected function resolveParameters(array $parameters): array
```

Comment:
Is this method intended to be used by the application's App instance? If not, it would be better to have it private instead to not expose any unnecessary complexity.

Reply:
This could also be made private, but there is a possible use case for extending it slightly and making it public.

Some frameworks allow injection of dependencies into controller action methods, which could be achieved using this method. It is however possible that that would just encourage bad habits in the form of overly large controllers.